### PR TITLE
New classes for commands and instructions

### PR DESCRIPTION
**Pull request by benjvmin93**
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `pytest`)
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
- Format added code by `black` and `isort`
  - See `pyproject.toml` for configurations

Then, please fill in below:

**Context (if applicable):**

**Description of the change:**
Introduced `class Command` and `class Instruction` for readability, maintainability and data validation purposes.

These classes allow to get rid of the list representation of both circuit model instructions, and the pattern commands. They are implemented using pydantic, which allows to validate the datas passed to the classes during instanciation.

**Related issue:**

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,14 +5,27 @@ All notable changes to this project will be documented in this file.
...
+
+- Added classes for the pattern commands and circuit instructions. It has been done for data validation, readability and maintainability purposes. All the commands and instructions were represented in raw inside lists which allowed wrong datas passed, but also made it difficult to turn the code into a more general way to build commands and instructions.
+  - Added `class Command` and all its child classes that represent all the pattern commands.
+  - Added `class Instruction` for the quantum circuit model. Every instruction can be instanciated using this class by passing its name as defined in the Enum `InstructionName`.
+
+### Fixed
+
+### Changed
+
+- Made all the updates inside the files making use of the commands and instructions list representation.
```
Comment (thierry-martinez) [inline CHANGELOG.md:27]: You should rephrase the change in terms of API changes: the user cannot write `pattern.add(["M", 0, "XY", 0, [], []])` anymore but should write `pattern.add(M(node=0))` for instance.

File: benchmarks/statevec.py [hunk 2]
```diff
@@ -99,7 +99,9 @@ def simple_random_circuit(nqubit, depth):
 # to transpile into a measurement pattern.
 
 
-def translate_graphix_rc_into_paddle_quantum_circuit(graphix_circuit: Circuit) -> PaddleCircuit:
+def translate_graphix_rc_into_paddle_quantum_circuit(
+    graphix_circuit: Circuit,
+) -> PaddleCircuit:
```
Comment (thierry-martinez) [inline benchmarks/statevec.py:104]: This kind of changes are probably due to `black` not having been called with `-l 120`. They should not appear in the PR.

File: examples/noisy_mbqc.py [hunk 3]
```diff
@@ -18,7 +18,7 @@
 theta = np.random.rand(2)
 circuit.rz(0, theta[0])
 circuit.rz(1, theta[1])
-circuit.cnot(0, 1)
+circuit.cnot(0,1)
```
Comment (thierry-martinez) [inline examples/noisy_mbqc.py:21]: This line should not have been changed.

Comment (thierry-martinez) [reply examples/noisy_mbqc.py:21]: More generally, all the changes in this file seem to be only reformatting: maybe we should reformat this file, but it should be done in another PR.

File: examples/pattern_fragments.py [hunk 4]
```diff
@@ -19,10 +19,10 @@
 pattern = circuit.transpile().pattern
 pattern.draw_graph(flow_from_pattern=False)
 
-# %%
+#%%
```
Comment (thierry-martinez) [inline examples/pattern_fragments.py:22]: This file should not have been changed.

File: examples/pattern_fragments.py [hunk 5]
```diff
@@ -19,10 +19,10 @@
...
 # Using :code:`opt=True` option for :code:`transpile` method, we switch to patterns with non-XY plane measurements allowed,
 # which has gflow (not flow). For CCX gate, the number of ancilla qubits required is nearly halved:
-pattern = circuit.transpile(opt=True).pattern
+pattern = circuit.transpile(opt=True)
```
Comment (thierry-martinez) [inline examples/pattern_fragments.py:25]: We should keep `.pattern`.

File: graphix/pattern.py [hunk 6]
```diff
@@ -66,11 +68,11 @@ def __init__(self, input_nodes=[]):
         self.__Nnode = len(input_nodes)  # total number of nodes in the graph state
         self._pauli_preprocessed = False  # flag for `measure_pauli` preprocessing completion
 
-        self.__seq = []
+        self.__seq: List[command.Command] = []
```
Comment (thierry-martinez) [inline graphix/pattern.py:71]: If you add `from __future__ import annotations`, you may use `list[command.Command]`.

File: graphix/pattern.py [hunk 7]
```diff
@@ -194,14 +194,14 @@ def __repr__(self):
             f"graphix.pattern.Pattern object with {len(self.__seq)} commands and {len(self.output_nodes)} output qubits"
         )
 
-    def equal(self, other):
+    def equal(self, other: "Pattern"):
```
Comment (thierry-martinez) [inline graphix/pattern.py:197]: If you use `from __future__ import annotations`, you may get rid of quotation marks around `Pattern`.

File: graphix/pattern.py [hunk 8]
```diff
@@ -194,14 +194,14 @@ def __repr__(self):
...
         return (
             self.__seq == other.__seq
             and self.input_nodes == other.input_nodes
             and self.output_nodes == other.output_nodes
         )
 
-    def print_pattern(self, lim=40, filter=None):
+    def print_pattern(self, lim=40, filter: List[str] = None):
```
Comment (thierry-martinez) [inline graphix/pattern.py:204]: It would be nicer to have `filter: list[command.CommandKind]`.

File: graphix/pattern.py [hunk 9]
```diff
@@ -400,26 +424,39 @@ def shift_signals(self, method="local"):
...
+                #     case command.CommandKind.S:
+                #         self._commute_SS(target)
+                #     case _:
+
                 target += 1
         else:
             raise ValueError("Invalid method")
 
-    def _find_op_to_be_moved(self, op, rev=False, skipnum=0):
+    def _find_op_to_be_moved(self, op: str, rev=False, skipnum=0):
```
Comment (thierry-martinez) [inline graphix/pattern.py:459]: May we have `op: command.CommandKind`?

File: graphix/command.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
+"""Data validator command classes."""
+
+from pydantic import BaseModel
+from typing import Union, Literal, List, Tuple
+import enum
+
+Node = int
+Plane = Union[Literal["XY"], Literal["YZ"], Literal["XZ"]]
```
Comment (thierry-martinez) [inline graphix/command.py:8]: Can be written `Literal["XY", "YZ", "XZ"]`. But may we use `pauli.Plane` instead?

**Review by thierry-martinez — CHANGES_REQUESTED** [review]
Well done! The diff set is huge but I believe it is mainly the fault of `black` having been invoked without `-l 120`. I suggested some corrections to improve even more the use of declarative types.

File: examples/noisy_mbqc.py (hunk 3 above)
Comment (benjvmin93) [reply examples/noisy_mbqc.py:21]: Is it really useful to make a whole PR for reformatting? I could maybe just revert the changes implied by the wrong use of black for the concerned files?

File: graphix/pattern.py (hunk 6 above)
Comment (benjvmin93) [reply graphix/pattern.py:71]: I guess I will also have to echo this change to the other `list` type specifications, along with the tuple ones?

**Comment by shinich1** [comment]
@thierry-martinez could you resolve the above comments if think they're already taken care of (if not, please leave them open)?

File: graphix/gflow.py
Comment (shinich1) [inline graphix/gflow.py:1]: could you update `get_corrections_from_pattern`?

File: graphix/simulator.py [hunk 11]
```diff
@@ -88,53 +90,94 @@ def run(self):
...
+                #         self.backend.correct_byproduct(cmd)
+                #         if np.mod(np.sum([self.results[j] for j in cmd.domain]), 2) == 1:
+                #             self.backend.apply_channel(self.noise_model.byproduct_z(), [cmd.node])
+                #     case CommandKind.C:
+                #         self.backend.apply_clifford(cmd.node)
+                #         self.backend.apply_channel(self.noise_model.clifford(), [cmd.node])
+                #     case CommandKind.T:
+                #         self.noise_model.tick_clock()
+                #     case _:
+                #         raise ValueError("Invalid commands.")
```
Comment (shinich1) [inline graphix/simulator.py:180]: let us wait until we deprecate (unsupport) <=py 3.9 to change to this.

File: graphix/transpiler.py [hunk 12]
```diff
@@ -763,33 +948,60 @@ def _move_byproduct_to_right(self):
...
+            #         self._commute_with_Rx(target)
+            #     case instruction.InstructionKind.RY:
+            #         self._commute_with_Ry(target)
+            #     case instruction.InstructionKind.RZ:
+            #         self._commute_with_Rz(target)
+            #     case instruction.InstructionKind.RZZ:
+            #         self._commute_with_Rzz(target)
+            #     case _:
+            #         # Pauli gates commute up to global phase.
+            #         self._commute_with_following(target)
```
Comment (shinich1) [inline graphix/transpiler.py:978]: same as a above

File: graphix/visualization.py [hunk 13]
```diff
@@ -299,7 +299,14 @@ def visualize_w_flow(
...
             if len(arrow_path[arrow]) == 2:
-                nx.draw_networkx_edges(self.G, pos, edgelist=[arrow], edge_color="black", arrowstyle="->", arrows=True)
+                nx.draw_networkx_edges(
+                    self.G,
+                    pos,
+                    edgelist=[arrow],
+                    edge_color="black",
+                    arrowstyle="->",
+                    arrows=True,
+                )
```
Comment (shinich1) [inline graphix/visualization.py:309]: would `black` with `-l 120` option help to minimize diff?

File: graphix/command.py
Comment (shinich1) [inline graphix/command.py:1]: let us add to API reference: in `docs/source/generator.rst`, under the heading `Pattern Generation`, add like the following:

```
:mod:`graphix.command` module
++++++++++++++++++++++++++++++++

.. currentmodule:: graphix.command

.. autoclass:: N

[add others]
```

File: graphix/command.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    M = "M"
+    E = "E"
+    C = "C"
+    X = "X"
+    Z = "Z"
+    T = "T"
+    S = "S"
+
+
+class Command(BaseModel):
```
Comment (shinich1) [inline graphix/command.py:21]: make this abc

File: graphix/instruction.py [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    Y = "Y"
+    Z = "Z"
+    I = "I"
+    M = "M"
+    RX = "RX"
+    RY = "RY"
+    RZ = "RZ"
+
+
+class Instruction(BaseModel):
```
Comment (shinich1) [inline graphix/instruction.py:27]: make this abc

File: graphix/command.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
+"""Data validator command classes."""
+
+from pydantic import BaseModel
+from typing import Union, Literal, List, Tuple
+from graphix.pauli import Plane
+import enum
+
+Node = int
+
+class CommandKind(str, enum.Enum):
```
Comment (shinich1) [inline graphix/command.py:10]: (comment by @EarlMilktea)  - it does not make sense for this to inherit from `str` - e.g.  this should not inherit [string methods](https://docs.python.org/3/library/stdtypes.html#string-methods).

File: graphix/sim/statevec.py [hunk 17]
```diff
@@ -201,7 +201,7 @@ def __init__(self, nqubit=1, plus_states=True):
     def __repr__(self):
         return f"Statevec, data={self.psi}, shape={self.dims()}"
 
-    def evolve_single(self, op, i):
+    def evolve_single(self, op: np.ndarray, i: int):
```
Comment (shinich1) [inline graphix/sim/statevec.py:204]: please annotate with `npt.NDArray` instead of `np.ndarray`.

**Comment by shinich1** [comment]
great work @benjvmin93 and sorry for the delay reviewing this. I added various comments. 
Let us aim to merge this after #135 is merged (should be very soon).

File: graphix/clifford.py [hunk 18]
```diff
@@ -97,30 +97,630 @@
...
-        [15, 7, 13, 14, 23, 20, 16, 0, 19, 21, 22, 18, 17, 3, 2, 1, 8, 11, 12, 6, 10, 4, 5, 9],
-        [16, 17, 18, 19, 6, 8, 15, 10, 14, 11, 12, 13, 7, 9, 5, 4, 20, 21, 22, 23, 0, 1, 2, 3],
-        [17, 16, 19, 18, 11, 12, 14, 4, 15, 6, 8, 7, 13, 5, 9, 10, 23, 22, 21, 20, 2, 3, 0, 1],
-        [18, 19, 16, 17, 12, 11, 13, 9, 7, 8, 6, 15, 14, 10, 4, 5, 21, 20, 23, 22, 3, 2, 1, 0],
-        [19, 18, 17, 16, 8, 6, 7, 5, 13, 12, 11, 14, 15, 4, 10, 9, 22, 23, 20, 21, 1, 0, 3, 2],
-        [20, 21, 22, 23, 15, 14, 4, 12, 5, 13, 7, 9, 10, 11, 8, 6, 0, 1, 2, 3, 16, 17, 18, 19],
-        [21, 20, 23, 22, 13, 7, 5, 6, 4, 15, 14, 10, 9, 8, 11, 12, 3, 2, 1, 0, 18, 19, 16, 17],
-        [22, 23, 20, 21, 7, 13, 9, 11, 10, 14, 15, 4, 5, 12, 6, 8, 1, 0, 3, 2, 19, 18, 17, 16],
-        [23, 22, 21, 20, 14, 15, 10, 8, 9, 7, 13, 5, 4, 6, 12, 11, 2, 3, 0, 1, 17, 16, 19, 18],
+        [
```
Comment (EarlMilktea) [inline graphix/clifford.py:100]: I feel it's not readable.

How about removing trailing commas?

File: graphix/sim/statevec.py [hunk 19]
```diff
@@ -1,13 +1,16 @@
 from copy import deepcopy
+from typing import List, Tuple, Union
```
Comment (EarlMilktea) [inline graphix/sim/statevec.py:2]: Please use `__futures__` instead.

File: graphix/sim/tensornet.py [hunk 20]
```diff
@@ -1,12 +1,15 @@
 import string
 from copy import deepcopy
+from typing import Union, List
```
Comment (EarlMilktea) [inline graphix/sim/tensornet.py:3]: Use `__future__`.

File: graphix/transpiler.py [hunk 21]
```diff
@@ -8,15 +8,43 @@
 
 import dataclasses
 from copy import deepcopy
-from typing import Optional, Sequence
-
+from typing import Optional, Sequence, List, Tuple
```
Comment (EarlMilktea) [inline graphix/transpiler.py:11]: Use `__future__`.

File: graphix/transpiler.py [hunk 22]
```diff
@@ -8,15 +8,43 @@
...
 import graphix.sim.base_backend
 from graphix.ops import Ops
 from graphix.pattern import Pattern
 from graphix.sim.statevec import Statevec
+from graphix import command
+from graphix.command import N, M, E, X, Z, CommandKind
+from graphix import instruction
+
+
+@dataclasses.dataclass
```
Comment (EarlMilktea) [inline graphix/transpiler.py:24]: Let's use either `pydantic.BaseModel` or `dataclass`, not both.

They have many in common.

File: graphix/transpiler.py (hunk 21 above)
Comment (EarlMilktea) [inline graphix/transpiler.py:11]: Please import from `collections.abc` rather than `typing`.

File: tests/test_pattern.py [hunk 23]
```diff
@@ -312,7 +313,7 @@ def swap(circuit: Circuit, a: int, b: int) -> None:
     circuit.cnot(a, b)
 
 
-class TestLocalPattern:
+class TestLocalPattern():
```
Comment (EarlMilktea) [inline tests/test_pattern.py:316]: Unnecessary `()`.

**Review by EarlMilktea — CHANGES_REQUESTED** [review]

File: graphix/pattern.py [hunk 24]
```diff
@@ -361,21 +393,21 @@ def is_standard(self):
...
+            while kind == command.CommandKind.E:
+                kind = next(it).kind
+            while kind == command.CommandKind.M:
+                kind = next(it).kind
+            xzc = {command.CommandKind.X, command.CommandKind.Z, command.CommandKind.C}
+            while kind in xzc:
+                kind = next(it).kind
+            return False
+        except StopIteration:
+            return True
```
Comment (shinich1) [inline graphix/pattern.py:382]: NOTE: the internal logic of `is_standard` has changed for performnce.

File: graphix/pattern.py [hunk 25]
```diff
@@ -637,67 +673,76 @@ def _move_N_to_left(self):
...
+                    cmd = self.__seq[index_Z + 1]
+                    if cmd.kind == command.CommandKind.M:
+                        search = self._commute_MZ(index_Z)
+                        if search:
+                            Z_limit -= 1  # ZM commutation rule removes Z command
+                            break
+                    else:
+                        self._commute_with_following(index_Z)
+                    index_Z += 1
+            index -= 1
```
Comment (shinich1) [inline graphix/pattern.py:692]: NOTE: also logic changed for performance.

**Review by shinich1 — APPROVED** [review]
All comments are resolved, and I believe the changes look sound. 
A few changes (optimizations) of the internal logics were marked above for future reference.
I ask @EarlMilktea to take a quick final look before we merge this.
Thank you @benjvmin93 and @thierry-martinez for the great work!

**Comment by EarlMilktea** [comment]
@benjvmin93 @thierry-martinez 

Great job!

I will submit small PRs directly to your fork later.

**Comment by EarlMilktea** [comment]
MEMO: Usage of `abc.ABC` may need refactor later.

**Review by EarlMilktea — APPROVED** [review]

**Review by thierry-martinez — APPROVED** [review]

**Comment by thierry-martinez** [comment]
I think it is all green now: @shinich1, what do you think? Could @benjvmin93 merge the PR?

**Comment by shinich1** [comment]
@benjvmin93 @thierry-martinez great, please squash and merge.

**Pull request merged**